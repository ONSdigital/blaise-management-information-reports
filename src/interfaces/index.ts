interface InterviewerCallHistoryReportData {
    questionnaire_name: string
    serial_number: string
    call_start_time: string
    dial_secs: string
    call_result: string
}

interface AppointmentResourcePlanningReportData {
    questionnaire_name: string
    appointment_time: string
    appointment_language: string
    total: number
}

export type {InterviewerCallHistoryReportData, AppointmentResourcePlanningReportData};
