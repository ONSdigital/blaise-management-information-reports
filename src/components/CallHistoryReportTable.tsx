import { InterviewerCallHistoryReport } from "../interfaces";
import { formatDateAndTime } from "../utilities/DateFormatter";
import { convertSecondsToMinutesAndSeconds } from "../utilities/Converters";
import React, { ReactElement } from "react";
import { ONSPanel } from "blaise-design-system-react-components";

interface CallHistoryReportTableProps {
    reportData: any
    messageNoData: string
}

function isEmpty(reportData: any): boolean {
    return !(reportData && reportData.length > 0);
}

export default function callHistoryReportTable({ reportData, messageNoData }: CallHistoryReportTableProps): ReactElement {
    if (isEmpty(reportData)) {
        return <ONSPanel hidden={messageNoData === "" && true}>{messageNoData}</ONSPanel>;
    }

    return (
        <>
            <table id="report-table" className="table u-mt-s">
                <thead className="table__head u-mt-m">
                    <tr className="table__row">
                        <th scope="col" className="table__header ">
                            <span>Questionnaire</span>
                        </th>
                        <th scope="col" className="table__header ">
                            <span>Serial Number</span>
                        </th>
                        <th scope="col" className="table__header ">
                            <span>Call Start Time</span>
                        </th>
                        <th scope="col" className="table__header ">
                            <span>Call Length</span>
                        </th>
                        <th scope="col" className="table__header ">
                            <span>Call Result</span>
                        </th>
                        <th scope="col" className="table__header ">
                            <span>Outcome Code</span>
                        </th>
                    </tr>
                </thead>
                <tbody className="table__body">
                    {
                        reportData.map((callHistory: InterviewerCallHistoryReport, index: number) => {
                            return (
                                <tr className="table__row" key={index}
                                    data-testid={"report-table-row"}>
                                    <td className="table__cell ">
                                        {callHistory.questionnaire_name}
                                    </td>
                                    <td className="table__cell ">
                                        {callHistory.serial_number}
                                    </td>
                                    <td className="table__cell ">
                                        {formatDateAndTime(callHistory.call_start_time)}
                                    </td>
                                    <td className="table__cell ">
                                        {convertSecondsToMinutesAndSeconds(callHistory.dial_secs)}
                                    </td>
                                    <td className="table__cell ">
                                        {(callHistory.call_result === null ? "Unknown" : callHistory.call_result)}
                                    </td>
                                    <td className="table__cell ">
                                        {callHistory.outcome_code}
                                    </td>
                                </tr>
                            );
                        })
                    }
                </tbody>
            </table>
        </>
    );
}
