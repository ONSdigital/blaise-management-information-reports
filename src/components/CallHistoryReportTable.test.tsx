import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import React from "react";
import { screen } from "@testing-library/dom";
import CallHistoryReportTable from "./CallHistoryReportTable";
import { InterviewerCallHistoryReport } from "../interfaces";

const reportData: InterviewerCallHistoryReport = {
    call_result: "Appointment",
    dial_secs: 90,
    questionnaire_name: "LMS2101_AA1",
    serial_number: "900001",
    call_start_time: "2022-01-31T09:00-0000",
    outcome_code: "320",
};
const messageNoData = "No data found for parameters given.";

describe("CallHistoryReportTable", () => {
    it("matches snapshot", async () => {
        const wrapper = render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(await wrapper).toMatchSnapshot();
    });

    it("displays a heading of 'Questionnaire'", async () => {
        render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(screen.queryByText("Questionnaire")).toBeVisible();
    });

    it("displays a heading of 'Serial number'", async () => {
        render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(screen.queryByText("Serial Number")).toBeVisible();
    });

    it("displays a heading of 'Call Start Time'", async () => {
        render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(screen.queryByText("Call Start Time")).toBeVisible();
    });

    it("displays a heading of 'Call Length'", async () => {
        render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(screen.queryByText("Call Length")).toBeVisible();
    });

    it("displays a heading of 'Call Result'", async () => {
        render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(screen.queryByText("Call Result")).toBeVisible();
    });

    it("displays a heading of 'Outcome Code'", async () => {
        render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(screen.queryByText("Outcome Code")).toBeVisible();
    });

    it("displays the relevant information for each heading", async () => {
        render(
            <CallHistoryReportTable messageNoData={messageNoData} reportData={[reportData]} />,
        );
        expect(screen.queryByText("LMS2101_AA1")).toBeVisible();
        expect(screen.queryByText("900001")).toBeVisible();
        expect(screen.queryByText("31/01/2022 09:00:00")).toBeVisible();
        expect(screen.queryByText("01:30")).toBeVisible();
        expect(screen.queryByText("Appointment")).toBeVisible();
        expect(screen.queryByText("320")).toBeVisible();
    });
});
