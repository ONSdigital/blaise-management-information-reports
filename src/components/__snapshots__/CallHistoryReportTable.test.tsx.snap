// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallHistoryReportTable matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="table u-mt-s"
        id="report-table"
      >
        <thead
          class="table__head u-mt-m"
        >
          <tr
            class="table__row"
          >
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Questionnaire
              </span>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Serial Number
              </span>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Call Start Time
              </span>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Call Length
              </span>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Call Result
              </span>
            </th>
            <th
              class="table__header "
              scope="col"
            >
              <span>
                Outcome Code
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          class="table__body"
        >
          <tr
            class="table__row"
            data-testid="report-table-row"
          >
            <td
              class="table__cell "
            >
              LMS2101_AA1
            </td>
            <td
              class="table__cell "
            >
              900001
            </td>
            <td
              class="table__cell "
            >
              31/01/2022 09:00:00
            </td>
            <td
              class="table__cell "
            >
              01:30
            </td>
            <td
              class="table__cell "
            >
              Appointment
            </td>
            <td
              class="table__cell "
            />
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="table u-mt-s"
      id="report-table"
    >
      <thead
        class="table__head u-mt-m"
      >
        <tr
          class="table__row"
        >
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Questionnaire
            </span>
          </th>
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Serial Number
            </span>
          </th>
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Call Start Time
            </span>
          </th>
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Call Length
            </span>
          </th>
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Call Result
            </span>
          </th>
          <th
            class="table__header "
            scope="col"
          >
            <span>
              Outcome Code
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        class="table__body"
      >
        <tr
          class="table__row"
          data-testid="report-table-row"
        >
          <td
            class="table__cell "
          >
            LMS2101_AA1
          </td>
          <td
            class="table__cell "
          >
            900001
          </td>
          <td
            class="table__cell "
          >
            31/01/2022 09:00:00
          </td>
          <td
            class="table__cell "
          >
            01:30
          </td>
          <td
            class="table__cell "
          >
            Appointment
          </td>
          <td
            class="table__cell "
          />
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
